pkg/intercp/catalog/writer.go

```diff
@@ -0,0 +1,64 @@
+package catalog
+
+import (
+	"context"
+	"sort"
+	"time"
+
+	"github.com/kumahq/kuma/pkg/core"
+	"github.com/kumahq/kuma/pkg/core/runtime/component"
+	"github.com/kumahq/kuma/pkg/core/user"
+)
+
+var writerLog = core.Log.WithName("intercp").WithName("catalog").WithName("writer")
+
+type catalogWriter struct {
+	catalog    Catalog
+	heartbeats *Heartbeats
+	instance   Instance
+	interval   time.Duration
+}
+
+var _ component.Component = &catalogWriter{}
+
+func NewWriter(catalog Catalog, heartbeats *Heartbeats, instance Instance, interval time.Duration) component.Component {
+	return &catalogWriter{
+		catalog:    catalog,
+		heartbeats: heartbeats,
+		instance:   instance,
+		interval:   interval,
+	}
+}
+
+func (r *catalogWriter) Start(stop <-chan struct{}) error {
+	heartbeatLog.Info("starting catalog writer")
+	ctx := user.Ctx(context.Background(), user.ControlPlane)
+	ticker := time.NewTicker(r.interval)
+	for {
+		select {
+		case <-ticker.C:
+			instances := r.heartbeats.ResetAndCollect()
+			r.instance.Leader = true
+			instances = append(instances, r.instance)
+			sort.Slice(instances, func(i, j int) bool {
+				return instances[i].Id < instances[j].Id
+			})
+			updated, err := r.catalog.Replace(ctx, instances)
```

Comment:
When you kill the leader and a new one is starting it's guaranteed to have `KUMA_INTER_CP_CATALOG_WRITER_INTERVAL` of a downtime. This is the time when the catalog is going to have only 1 entry of a leader. 

What if the catalog will have a structure:
```yaml
leader: {}
instances: []
```
when the new leader is starting it immediately updates the `leader` field and updates `instances` after `KUMA_INTER_CP_CATALOG_WRITER_INTERVAL`?

Reply:
done using replacement

Reply:
Yep, this works as well